### PR TITLE
COCOS-103 - User Identification and Auth using digital signatures

### LINKED ISSUE #103: Feature: Enforce User Roles and Identity
### Is your feature request related to a problem? Please describe.

- Agent is currently unable to distinguish different users by their roles ie (algo providers, dataset providers, result consumers)

### Describe the feature you are requesting, as well as the possible use case(s) for it.

- Users should be identified by their certs(mTLS) for different roles by agent

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by SammyOina**
# What type of PR is this?
This a feature as it introduces digital signatures for user authorization and identification

## What does this do?
image

- **New Features**
  - Refactored `agentSDK` struct in `pkg/sdk/agent.go` to implement the `SDK` interface with new methods for handling algorithms, data, results, and attestation.
  - Introduced functionality in `cli/keys.go` to generate RSA public/private key pairs.
  - Added private key handling in `cli/datasets.go` and `cli/result.go` for secure operations.
  - Introduced authentication and authorization features in the service with the `auth.go` file.

- **Updates**
  - Enhanced authentication and authorization features in the service with adjustments to the `auth.go` file.
  - Updated Go version to `1.22.x` in all workflows.
  - Modified the `ResultConsumers` field in `test/manual/agent-config/main.go` to enhance representation.
  - Modified the `ComputationRunReq` message in `manager/manager.proto` to use `ResultConsumer` message and added `userKey` fields to `Dataset` and `Algorithm` messages.

- **Bug Fixes**
  - Adjusted command line argument handling for better usability and security in `test/manual/README.md`.

- **Chores**
  - Added `*.pem` to `.gitignore` for better security practices.

- **Documentation**
  - Updated test instructions in `test/manual/README.md` to reflect new argument order for `algorithm-file` command.

## Which issue(s) does this PR fix/relate to?

- Resolves #103

## Have you included tests for your changes?
No

## Did you document any new/modified feature?
Yes

### CODE REVIEW

File: agent/auth/auth.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		}
+		md, ok := metadata.FromIncomingContext(stream.Context())
+		if !ok {
+			return status.Errorf(codes.Unauthenticated, "missing metadata")
+		}
+		userID, signature, err := extractSignatureAndUserID(md)
+		if err != nil {
+			return status.Errorf(codes.Unauthenticated, "invalid metadata")
+		}
+
```
Comment (drasko) [inline agent/auth/auth.go:115]: If you added a white line here for readability, add it consistently to the upper ifs as well

File: cli/keys.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			defer privFile.Close()
+
+			if err := pem.Encode(privFile, &pem.Block{
+				Type:  privateKeyType,
+				Bytes: x509.MarshalPKCS1PrivateKey(privKey),
+			}); err != nil {
+				log.Fatalf("Error encoding private key: %v", err)
+			}
+
+			pubFile, err := os.Create("public.pem")
```
Comment (drasko) [inline cli/keys.go:50]: This name of the file is a string constant, so define it before

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Comment by drasko** [comment]
@SammyOina CI is falling, please fix

**Review by drasko — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]
Auth layer should be moved from the API closer to svc (but you don't have to move it to the service exactly), and we should have some abstractions in the form of interfaces so we can more easily test it and decouple implementation from specification.

**Pull request merged**